test/e2e/horizontal_pod_autoscaling.go

```diff
@@ -79,17 +95,83 @@ type HPAScaleTest struct {
 // The first state change is due to the CPU being consumed initially, which HPA responds to by changing pod counts.
 // The second state change is due to the CPU burst parameter, which HPA again responds to.
 // TODO The use of 3 states is arbitrary, we could eventually make this test handle "n" states once this test stabilizes.
-func (scaleTest *HPAScaleTest) run(name, kind string, rc *ResourceConsumer, f *Framework) {
-	rc = NewDynamicResourceConsumer(name, kind, scaleTest.initPods, scaleTest.cpuStart, 0, scaleTest.maxCPU, 100, f)
+func (scaleTest *HPAScaleTest) run(name, kind string, f *Framework) bool {
+	rc := NewDynamicResourceConsumer(name, kind, scaleTest.initPods, scaleTest.cpuStart, 0, scaleTest.maxCPU, 100, f)
 	defer rc.CleanUp()
 	createCPUHorizontalPodAutoscaler(rc, scaleTest.idealCPU, scaleTest.minPods, scaleTest.maxPods)
-	rc.WaitForReplicas(scaleTest.firstScale)
+	passed := rc.WaitForReplicas(scaleTest.firstScale)
+	if !passed {
+		Logf("first scale: timeout waiting for pods size to be %d", scaleTest.secondScale)
+		return false
+	}
+	// Apparently this is failing, because deleted.  logging time to see why.
 	rc.EnsureDesiredReplicas(scaleTest.firstScale, scaleTest.firstScaleStasis)
 	rc.ConsumeCPU(scaleTest.cpuBurst)
-	rc.WaitForReplicas(scaleTest.secondScale)
+	passed = rc.WaitForReplicas(scaleTest.secondScale)
+	if !passed {
+		Logf("second scale: timeout waiting for pods size to be %d", scaleTest.secondScale)
+		return false
+	}
+	return true
+}
+
+// Tests that multiple HPAs work properly, in parallel.
+// Right now, we only test 2 generations.
+func scaleMultitest(name, kind string, f *Framework) {
+	testResults := [2]bool{}
+	testDone := [2]bool{}
+	Logf("start %v", testResults)
+
+	// Important: make sure exactly len(testResults) go func's are created,
+	// and that each one writes to a unique element of the tests array.
+	for i := 0; i < len(testResults); i += 2 {
+		go func(ii int) {
+			defer GinkgoRecover()
+			testResults[ii] = scaleUp(fmt.Sprint(name, "up", ii), kind, f)
+			testDone[ii] = true
+		}(i)
+		go func(ii int) {
+			defer GinkgoRecover()
+			testResults[ii] = scaleDown(fmt.Sprint(name, "down", ii), kind, f)
```

Comment:
a little confused here - don't we need to make sure all of the `scaleUp` calls are done before `scaleDown`?